### PR TITLE
Fix/oracle bug with token decimal handling

### LINKED ISSUE #25: Bug: LP token underpriced when token decimals are identical & below 18.
PoC:

```
    function test_Bug_AdjustDecimals() external whenTokenDecimalsSame whenTokenDecimalsLessThan18 {
        // Both tokens have 6 decimals
        reinitOracle(6, 6);

        // Setup
        int256 answer0 = 1e8;
        int256 answer1 = 1e8;
        uint256 token0PoolReserve = 10e6;
        uint256 token1PoolReserve = 11e6;

        setLatestRoundDataMocks(answer0, answer1, token0PoolReserve, token1PoolReserve, defaults.WEIGHT_50());

        (uint256 price0, uint256 price1, uint256 updatedAt_) = oracle.exposed_getFeedData();
        GPv2Order.Data memory order = oracle.exposed_simulateOrder(price0, price1);
        (uint256 token0Bal, uint256 token1Bal) = oracle.exposed_simulatePoolReserves(order);
        // IERC20(POOL).totalSupply() == 1000e18;

        // expected price:
        // (1 * 10.47 + 1 * 10.5 ) / 1000 ≈ 0.02

        // actual price:
        (, int256 answer,,,) = oracle.latestRoundData();
        assertEq(uint256(answer), 0);
    }
```

Need to refactor the decimals accounting logic in the contract.

### PR DESCRIPTION
**Pull request by lumoswiz**
Depends on #24  and should be merged after it.

Fixes #25 Corrects LP token pricing when underlying tokens have identical decimals & < 18.

Previous problem:
- Using `_adjustDecimals` on pool token balances when both tokens had identical decimals led to LP token underpricing.

Fix:
- Remove `_adjustDecimals` handling of pool token balances.
- Add `_calculatePrice` helper that:
  - Normalizes pool token liquidity values to 18 decimals, properly accounting for both token and feed decimals.
  - Calculates final LP token price in USD (8 decimals).
  - Assumes LP token implementation uses 18 decimals.

Additional improvements:
- Added explicit return value assignments across functions.

### CODE REVIEW

File: test/unit/LatestRoundData.t.sol [hunk 1]
```diff
@@ -219,40 +186,30 @@ contract LatestRoundData_Unit_Test is BaseTest {
...
-        uint256 tokenAmountIn =
-            calcInGivenOut(token0PoolReserve, WEIGHT_80, token1PoolReserve, WEIGHT_20, tokenAmountOut, 0);
+        uint256 tokenAmountOut = maxAmountOutGivenBalanceOut(1000e18);
+        uint256 tokenAmountIn = calcInGivenOut(1e18, 0.8e18, 1000e18, 0.2e18, tokenAmountOut, 0);
         token1PoolReserve -= tokenAmountOut;
         token0PoolReserve += tokenAmountIn;
 
-        // NaivePrice ≈ 10e8 == (0.5 * 4000 + 16000 * 1)
-        uint256 naivePrice = (token0PoolReserve * 4000e8 + token1PoolReserve * 1e8) / LP_TOKEN_SUPPLY;
+        // NaivePrice ≈ 18e8 == (0.5 * 4000 + 16000 * 1)
```
Comment (bh2smith) [inline test/unit/LatestRoundData.t.sol]: Good catch.

File: test/unit/fuzz/adjust-decimals/adjustDecimals.tree [hunk 2]
```diff
@@ -0,0 +1,11 @@
+adjustDecimals.tsol
```
Comment (bh2smith) [inline test/unit/fuzz/adjust-decimals/adjustDecimals.tree:1]:
```suggestion
adjustDecimals.t.sol
```

File: test/unit/fuzz/adjust-decimals/adjustDecimals.tree [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+├── when decimals are equal
+│   └── it should return values unchanged
+└── when decimals are not equal
+    ├── when at least one decimal is greater than 18
+    │   └── it should revert
+    └── when both decimals are less than, or equal to, 18
+        ├── when values overflow when adjusted
+        │   └── it should revert
+        └── when values can be adjusted without overflow
+            └── it should return the adjusted values with base 18 decimals
```
Comment (bh2smith) [inline test/unit/fuzz/adjust-decimals/adjustDecimals.tree:11]: Nit:
```suggestion
            └── it should return the adjusted values with base 18 decimals

```

File: test/unit/fuzz/calculate-price/calculatePrice.tree [hunk 4]
```diff
@@ -0,0 +1 @@
+calculatePrice.t.sol
```
Comment (bh2smith) [inline test/unit/fuzz/calculate-price/calculatePrice.tree:1]: I am not sure what these tree files are supposed to be about.
```suggestion
calculatePrice.t.sol

```

File: test/unit/fuzz/constructor/constuctor.tree [hunk 5]
```diff
@@ -0,0 +1,19 @@
+constructor.t.sol
```
Comment (bh2smith) [inline test/unit/fuzz/constructor/constuctor.tree:1]: The layout of this file is different from adjustDecimals.

**Review by bh2smith — CHANGES_REQUESTED** [review]
There is a lot to look at here. What are the tree files for? calculateUsd tree file is basically empty. There is some inconsistency between different tree file formatting.

Generally, I would prefer if the tests were migrated in a completely separate PR and the changes to the LPOracle contract addressing #25 elsewhere. Given that the contract changes are minimal it ok to leave here. However, I the diff is messed up because the order of adjustDecimals function was moved so it looks like a whole new function and I can't see what changed. I am unable to determine how the changes in this contract are directly addressing bug #25 (possibly because of the messed up diff).

Could you please move adjust decimals back to where it was so I can see what changed and maybe try to reduce the diff in the contract to only what is necessary? I'm uncomfortable approving so many arbitrary changes to the contract.

Shouldn't there also be a test introduced that demonstrates the Bug #25 is now resolved? If there is one, I can't tell where it is.

Screenshot 2024-12-19 at 19 55 07

Screenshot 2024-12-19 at 19 54 56

File: test/unit/fuzz/constructor/constuctor.tree (hunk 5 above)
Comment (lumoswiz) [reply test/unit/fuzz/constructor/constuctor.tree:1]: Forgot to convert this to a tree.

File: test/unit/fuzz/calculate-price/calculatePrice.tree (hunk 4 above)
Comment (lumoswiz) [reply test/unit/fuzz/calculate-price/calculatePrice.tree:1]: See [this](https://prberg.com/presentations/solidity-summit-2023/).

Examples: 
- [balancer cow-amm](https://github.com/balancer/cow-amm/tree/main/test/unit)
- [sablier v2](https://github.com/sablier-labs/v2-core)

**Comment by lumoswiz** [comment]
Based on this feedback I will be closing this PR & implement the changes in smaller increments.

**Pull request closed without merging**